instance/factories.py

```diff
@@ -0,0 +1,118 @@
+    """
+    Check environment and report potential problems for production instances
+    """
+    if not settings.SWIFT_ENABLE:
+        logger.warning("Swift support is currently disabled. Adjust SWIFT_ENABLE setting.")
+    if settings.INSTANCE_MYSQL_URL is None:
+        logger.warning("URL for external MySQL database is missing. Adjust INSTANCE_MYSQL_URL setting.")
+    if settings.INSTANCE_MONGO_URL is None:
+        logger.warning("URL for external Mongo database is missing. Adjust INSTANCE_MONGO_URL setting.")
+
+
+def instance_factory(**kwargs):
+    """
+    Factory function for creating instances.
+
+    Returns a newly created OpenEdXInstance.
+
+    Callers can use keyword arguments to pass in non-default values
+    for any field that is defined on the OpenEdXInstance model.
+
+    The only mandatory argument is `sub_domain`.
+
+    When called without any additional arguments, the instance that is returned
+    will have its fields set to default values that are appropriate
+    for *sandbox* instances.
+
+    To create an instance with default settings that are suitable for production,
+    use `production_instance_factory`.
+    """
+    # Ensure caller provided required arguments
+    assert "sub_domain" in kwargs
+
+    # Create instance
+    instance = OpenEdXInstance.objects.create(**kwargs)
+    return instance
+
+
+def production_instance_factory(**kwargs):
+    """
+    Factory function for creating production instances.
+
+    Returns a newly created OpenEdXInstance.
+
+    Callers can use keyword arguments to pass in non-default values
+    for any field that is defined on the OpenEdXInstance model.
+
+    The only mandatory argument is `sub_domain`.
+
+    When called without any additional arguments, the instance that is returned
+    will have its fields set to default values that are appropriate
+    for *production* instances.
+
+    To create an instance with default settings that are suitable for sandboxes,
+    use `instance_factory`.
+    """
+    # NOTE: The long-term goal is to eliminate differences between sandboxes
+    # and production instances, and for this function to disappear.
+    # Please do not add behavior that is specific to production instances here.
+
+    # Ensure caller provided required arguments
+    assert "sub_domain" in kwargs
+
+    # Check environment and report potential problems
+    _check_environment()
+
+    # Create instance
+    production_instance = OpenEdXInstance(**kwargs)
+    if "use_ephemeral_databases" not in kwargs:
+        production_instance.use_ephemeral_databases = False
+    if "configuration_version" not in kwargs:
+        production_instance.configuration_version = settings.LATEST_OPENEDX_RELEASE
+    if "openedx_release" not in kwargs:
+        production_instance.openedx_release = settings.LATEST_OPENEDX_RELEASE
+    if "configuration_extra_settings" not in kwargs:
+        template = loader.get_template('instance/ansible/prod-vars.yml')
```

Comment:
Instead of only loading the template when no `configuration_extra_settings` are provided, can we merge them using the merge function in `ansible.py`?  It would be convenient to provide more settings without losing the defaults.

I also suggest reversing the logic here to avoid the somewhat repetitive ifs:

    extra_settings = ... merge code using kwargs.pop("configuration_extra_settings") ...
    instance_kwargs = dict(
        use_ephemeral_databases=False,
        configuration_version=settings.LATEST_OPENEDX_RELEASE,
        openedx_release=settings.LATEST_OPENEDX_RELEASE,
        configuration_extra_settings=extra_settings,
    )
    instance_kwargs.update(kwargs)
    production_instance = OpenEdXInstance(**instance_kwargs)

(Though this is probably just a matter of personal taste.)